tools/cdb2sql/cdb2sql.cpp

```diff
@@ -2001,6 +2048,170 @@ static void int_handler(int signum)
     send_cancel_cnonce(cdb2_cnonce(cdb2h));
 }
 
+class InputProcessor {
+public:
+    InputProcessor() {}
+    virtual void process_input() = 0;
+
+    InputProcessor(const InputProcessor&) = delete;
+    InputProcessor(InputProcessor&&) = delete;
+    InputProcessor& operator=(const InputProcessor&) = delete;
+    InputProcessor& operator=(InputProcessor&&) = delete;
+
+};
+
+// Processes all statements as delimited blocks.
+// TODO: Put in its own file
+class BlockInputProcessor : public InputProcessor
+{
+public:
+    BlockInputProcessor() : _sql(Sql())
+    {
+        run_raw_statement("set multiline on", 0, NULL, 0, 0);
+    }
+
+    void process_input()
+    {
+        for(const char * line; (line = read_line(main_prompt)) != NULL;) {
+            process_line(line);
+        }
+        handle_trailing_sql();
+    }
+
+private:
+    class Sql {
+    public:
+        Sql() : _sql_str(std::string("")) {}
+
+        void trim_lhs_whitespace()
+        {
+            const std::string::iterator first_non_whitespace_char = std::find_if_not(
+                _sql_str.begin(), _sql_str.end(), ::isspace);
+
+            _sql_str.erase(_sql_str.begin(), first_non_whitespace_char);
+        }
+
+        void trim_rhs_whitespace()
+        {
+            const std::string::reverse_iterator last_non_whitespace_char = std::find_if_not(
+                _sql_str.rbegin(), _sql_str.rend(), ::isspace);
+
+            _sql_str.erase(last_non_whitespace_char.base());
+        }
+
+        int trim_lhs_stmt()
+        {
+            char *tailstr;
+            const int rc = cdb2_get_property(cdb2h, "sql:tail", &tailstr);
+            if (rc) { return 1; }
+
+            const int tailoff = atoi(tailstr);
+            free(tailstr);
+            _sql_str.erase(0, tailoff);
+
+            return 0;
+        }
+
+        const char * c_str()
+        {
+            return _sql_str.c_str();
+        }
+
+        void append_line(const char * const line)
+        {
+            _sql_str.append(line);
+            _sql_str.append("\n");
+        }
+
+        bool is_delimited()
+        {
+            return _sql_str.find(_delim) != std::string::npos;
+        }
+
+    private:
+        std::string _sql_str;
+    };
+
+    Sql _sql;
+    static const char _delim = ';';
+
+    void process_line(const char * const line)
+    {
+        if (is_comment_or_empty_line(_sql.c_str())
+            && is_comment_or_empty_line(line)) {
+            return;
```

Comment:
The input line is useless if it is a comment or an empty line and the preceding sql is a comment or an empty line